File: section-05-basic-hashing/03-sum-of-highest-and-lowest-frequency/sum-of-highest-and-lowest-frequency-opt.cs
public class Solution
{
    public int SumHighestAndLowestFrequency(int[] nums)
    {
        int maxFreq = 0,
            minFreq = nums.Length;
        Dictionary<int, int> map = new Dictionary<int, int>();
        for (int i = 0; i < nums.Length; i++)
        {
            if (map.ContainsKey(nums[i]))
                map[nums[i]]++;
            else
                map[nums[i]] = 1;
        }
        foreach (int freq in map.Values)
        {
            maxFreq = Math.Max(maxFreq, freq);
            minFreq = Math.Min(minFreq, freq);
        }
        return maxFreq + minFreq;
    }
}

/* 

Time Complexity 

* You loop through the array once to build the frequency map → O(n)
* Then you loop through the dictionary values → at most O(n) (if all elements are unique)
* No nested loops, everything is sequential
* So total time = O(n) + O(n)
* Final Time Complexity = O(n)✅

---

Space Complexity 

* You are storing frequencies in a dictionary
* In worst case (all elements different), dictionary stores n entries
* No extra space apart from this map
* So space used grows with input size
* Final Space Complexity = O(n)✅

 */

File: section-05-basic-hashing/03-sum-of-highest-and-lowest-frequency/sum-of-highest-and-lowest-frequency-opt.py
class Solution:
    # Function to get the sum of highest
    # and lowest frequency in array
    def sumHighestAndLowestFrequency(self, nums):
        
        # Variable to store maximum 
        # and minimum frequency
        maxFreq = 0
        minFreq = n
        
        # HashMap
        mpp = {}
        
        # Iterating on the array
        for num in nums:
            # Updating hashmap
            if num in mpp:
                mpp[num] += 1
            else:
                mpp[num] = 1
                
        # Iterate on the map
        for freq in mpp.values():
            # Update maximum and 
            # minimum frequencies
            maxFreq = max(maxFreq, freq)
            minFreq = min(minFreq, freq)
            
        # Return the required sum
        return maxFreq + minFreq

